File: Android/src/build-backups/kapt3-20260411-195127/stubs/debug/selfgemma/talk/data/roleplay/repository/AndroidRoleplayInteropDocumentRepository.java
package selfgemma.talk.data.roleplay.repository;

@javax.inject.Singleton()
@kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000<\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0010\u000e\n\u0002\b\u0003\n\u0002\u0010\u0002\n\u0002\b\u0003\n\u0002\u0010\u0012\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\b\u0007\u0018\u00002\u00020\u0001B\u0013\b\u0007\u0012\b\b\u0001\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\u0004\b\u0004\u0010\u0005J\u0016\u0010\u0006\u001a\u00020\u00072\u0006\u0010\b\u001a\u00020\u0007H\u0096@\u00a2\u0006\u0002\u0010\tJ\u001e\u0010\n\u001a\u00020\u000b2\u0006\u0010\b\u001a\u00020\u00072\u0006\u0010\f\u001a\u00020\u0007H\u0096@\u00a2\u0006\u0002\u0010\rJ\u0016\u0010\u000e\u001a\u00020\u000f2\u0006\u0010\b\u001a\u00020\u0007H\u0096@\u00a2\u0006\u0002\u0010\tJ\u001e\u0010\u0010\u001a\u00020\u000b2\u0006\u0010\b\u001a\u00020\u00072\u0006\u0010\f\u001a\u00020\u000fH\u0096@\u00a2\u0006\u0002\u0010\u0011J\u0016\u0010\u0012\u001a\u00020\u00132\u0006\u0010\b\u001a\u00020\u0007H\u0096@\u00a2\u0006\u0002\u0010\tJ\u0012\u0010\u0014\u001a\u0004\u0018\u00010\u00152\u0006\u0010\b\u001a\u00020\u0007H\u0002J\u0012\u0010\u0016\u001a\u0004\u0018\u00010\u00172\u0006\u0010\b\u001a\u00020\u0007H\u0002R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u0018"}, d2 = {"Lselfgemma/talk/data/roleplay/repository/AndroidRoleplayInteropDocumentRepository;", "Lselfgemma/talk/domain/roleplay/repository/RoleplayInteropDocumentRepository;", "context", "Landroid/content/Context;", "<init>", "(Landroid/content/Context;)V", "readText", "", "uri", "(Ljava/lang/String;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "writeText", "", "content", "(Ljava/lang/String;Ljava/lang/String;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "readBytes", "", "writeBytes", "(Ljava/lang/String;[BLkotlin/coroutines/Continuation;)Ljava/lang/Object;", "getMetadata", "Lselfgemma/talk/domain/roleplay/repository/RoleplayInteropDocumentMetadata;", "openInputStream", "Ljava/io/InputStream;", "openOutputStream", "Ljava/io/OutputStream;", "app_debug"})
public final class AndroidRoleplayInteropDocumentRepository implements selfgemma.talk.domain.roleplay.repository.RoleplayInteropDocumentRepository {
    @org.jetbrains.annotations.NotNull()
    private final android.content.Context context = null;
    
    @javax.inject.Inject()
    public AndroidRoleplayInteropDocumentRepository(@dagger.hilt.android.qualifiers.ApplicationContext()
    @org.jetbrains.annotations.NotNull()
    android.content.Context context) {
        super();
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object readText(@org.jetbrains.annotations.NotNull()
    java.lang.String uri, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super java.lang.String> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object writeText(@org.jetbrains.annotations.NotNull()
    java.lang.String uri, @org.jetbrains.annotations.NotNull()
    java.lang.String content, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object readBytes(@org.jetbrains.annotations.NotNull()
    java.lang.String uri, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super byte[]> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object writeBytes(@org.jetbrains.annotations.NotNull()
    java.lang.String uri, @org.jetbrains.annotations.NotNull()
    byte[] content, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object getMetadata(@org.jetbrains.annotations.NotNull()
    java.lang.String uri, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super selfgemma.talk.domain.roleplay.repository.RoleplayInteropDocumentMetadata> $completion) {
        return null;
    }
    
    private final java.io.InputStream openInputStream(java.lang.String uri) {
        return null;
    }
    
    private final java.io.OutputStream openOutputStream(java.lang.String uri) {
        return null;
    }
}

File: Android/src/app/src/main/java/selfgemma/talk/data/roleplay/repository/AndroidRoleplayInteropDocumentRepository.kt
package selfgemma.talk.data.roleplay.repository

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import dagger.hilt.android.qualifiers.ApplicationContext
import java.io.File
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import selfgemma.talk.domain.roleplay.repository.RoleplayInteropDocumentMetadata
import selfgemma.talk.domain.roleplay.repository.RoleplayInteropDocumentRepository

@Singleton
class AndroidRoleplayInteropDocumentRepository
@Inject
constructor(@ApplicationContext private val context: Context) : RoleplayInteropDocumentRepository {
  override suspend fun readText(uri: String): String =
    withContext(Dispatchers.IO) {
      readBytes(uri).decodeToString()
    }

  override suspend fun writeText(uri: String, content: String) {
    writeBytes(uri, content.toByteArray())
  }

  override suspend fun readBytes(uri: String): ByteArray =
    withContext(Dispatchers.IO) {
      openInputStream(uri)?.use { it.readBytes() } ?: error("Unable to open input stream for uri=$uri")
    }

  override suspend fun writeBytes(uri: String, content: ByteArray) {
    withContext(Dispatchers.IO) {
      openOutputStream(uri)?.use { output ->
        output.write(content)
      } ?: error("Unable to open output stream for uri=$uri")
    }
  }

  override suspend fun getMetadata(uri: String): RoleplayInteropDocumentMetadata =
    withContext(Dispatchers.IO) {
      val parsedUri = Uri.parse(uri)
      if (parsedUri.scheme.isNullOrBlank() || parsedUri.scheme == "file") {
        val file = if (parsedUri.scheme == "file") File(parsedUri.path.orEmpty()) else File(uri)
        return@withContext RoleplayInteropDocumentMetadata(
          displayName = file.name.takeIf { it.isNotBlank() },
          mimeType = context.contentResolver.getType(parsedUri),
        )
      }

      var displayName: String? = null
      context.contentResolver.query(parsedUri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)?.use { cursor ->
        if (cursor.moveToFirst()) {
          val index = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
          if (index >= 0) {
            displayName = cursor.getString(index)
          }
        }
      }

      RoleplayInteropDocumentMetadata(
        displayName = displayName,
        mimeType = context.contentResolver.getType(parsedUri),
      )
    }

  private fun openInputStream(uri: String) =
    run {
      val parsedUri = Uri.parse(uri)
      if (parsedUri.scheme.isNullOrBlank()) {
        File(uri).inputStream()
      } else if (parsedUri.scheme == "file") {
        File(parsedUri.path.orEmpty()).inputStream()
      } else {
        context.contentResolver.openInputStream(parsedUri)
      }
    }

  private fun openOutputStream(uri: String) =
    run {
      val parsedUri = Uri.parse(uri)
      if (parsedUri.scheme.isNullOrBlank()) {
        File(uri).outputStream()
      } else if (parsedUri.scheme == "file") {
        File(parsedUri.path.orEmpty()).outputStream()
      } else {
        context.contentResolver.openOutputStream(parsedUri, "w")
      }
    }
}
